### PR TITLE
@W-11500523@: Added pathless SFGE run to scanner:run command.

### PR DESCRIPTION
**Pull request by jfeingold35**
This PR does the following:
- Reorganizes several message files.
- Adds `--projectdir` and `--ignore-parse-errors` flag to `scanner:run`, equivalent to those on `scanner:run:dfa`
- `scanner:run` can now request SFGE as an engine.
    - If SFGE explicitly requested and `--projectdir` is missing, an error is thrown.
    - If SFGE is merely enabled, rather than specifically requested, and `--projectdir` is absent, SFGE isn't run, and a verbose warning is logged.
- `scanner:run` can now use SFGE to run non-DFA rules against files.
- The `UnusedMethodRule` rule has been enabled.

The following still needs to be done:
- Add example with `--projectdir` to `scanner:run` help text.
- Pathless SFGE needs to be added to smoke tests.

### CODE REVIEW

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 1]
```diff
@@ -120,97 +126,77 @@ export class SfgeEngine extends AbstractRuleEngine {
...
 	 * Helps make decision to run an engine or not based on the Rules, Target paths, and the Engine Options selected per
 	 * run. At this point, filtering should have already happened.
+	 * @override
 	 */
 	// eslint-disable-next-line @typescript-eslint/no-unused-vars
-	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string,string>): boolean {
-		// If the engine isn't filtered out, there's no reason to not run it.
-		return true;
+	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): boolean {
+		// If the EngineOptions doesn't include an SFGE config, don't try to run this engine.
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:133]: The new validation in `run.ts` prevents SFGE from being requested if `--projectdir` is absent. This is for if SFGE is merely default-enabled and the flag is missing. We just skip it and log a warning instead of halting with an error.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 2]
```diff
@@ -120,97 +126,77 @@ export class SfgeEngine extends AbstractRuleEngine {
...
 				targetCount += t.methods.length;
 			} else {
 				targetCount += t.paths.length;
 			}
 		});
 
+		// The rules have yet to be filtered by DFA/Non-DFA, so it's possible that some of the provided rules
+		// don't actually belong to this SFGE subvariant. So we should filter the rules so we only run ones
+		// in this engine's catalog.
+		const catalogRuleNames: Set<string> = new Set();
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:169]: Depending on the filter criteria, it's possible for the list of rules given to this method to contain both DFA and non-DFA rules. This lets us filter out the incompatible rules.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 3]
```diff
@@ -120,97 +126,77 @@ export class SfgeEngine extends AbstractRuleEngine {
...
-			this.logger.trace(`${SfgeEngine.ENGINE_NAME} evaluation failed. ${message}`);
-			throw new SfdxError(SfgeEngine.processStderr(message));
-		}
-	}
-
-	/**
-	 * TODO: Not supported yet. Idea is to detect if a custom rule path is supported by this engine.
-	 */
-	// eslint-disable-next-line @typescript-eslint/no-unused-vars
-	public matchPath(path: string): boolean {
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:170]: Method still here, just moved.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 4]
```diff
@@ -120,97 +126,77 @@ export class SfgeEngine extends AbstractRuleEngine {
...
-		// TODO: Implement this method for real, eventually.
-		return false;
-	}
-
-	/**
-	 * Returns value of `isEngineEnabled` based on Config or an internal decision.
-	 */
-	public async isEnabled(): Promise<boolean> {
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:178]: Method still here, just moved.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 5]
```diff
@@ -120,97 +126,77 @@ export class SfgeEngine extends AbstractRuleEngine {
...
-	public isEngineRequested(filterValues: string[], engineOptions: Map<string,string>): boolean {
-		// If the engine was specifically requested, or there were no engine filters at all, then this engine should be
-		// treated as requested. That way:
-		// - SFGE will be included in cataloging for `scanner:rule:list` if no filters are provided or if it's requested,
-		// - SFGE will be manually excluded from `scanner:run` by virtue of being a DFA engine,
-		// - SFGE will be manually included in `scanner:run:dfa` by virtue of being a DFA engine.
-		return EngineUtils.isFilterEmptyOrNameInFilter(this.getName(), filterValues);
-	}
-
-	public isDfaEngine(): boolean {
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:196]: Method now abstract.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 6]
```diff
@@ -243,25 +229,67 @@ export class SfgeEngine extends AbstractRuleEngine {
...
-				sourceType: sfgeViolation.sourceType,
-				sourceMethodName: sfgeViolation.sourceVertexName
-			});
+			result.violations.push(this.convertViolation(sfgeViolation));
 			resultMap.set(indexFile, result);
 		}
 		return [...resultMap.values()];
 	}
 
+	protected abstract convertViolation(sfgeViolation: SfgeViolation): RuleViolation;
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:238]: The subvariants have different ways of converting an SFGE violation into a real violation, so this is abstract.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 7]
```diff
@@ -120,97 +126,77 @@ export class SfgeEngine extends AbstractRuleEngine {
...
-		return EngineUtils.isFilterEmptyOrNameInFilter(this.getName(), filterValues);
-	}
-
-	public isDfaEngine(): boolean {
-		// NOTE: If SFGE implements and exposes non-DFA rules, then this will no longer be accurate. In that case, the
-		// best course of action is probably to divide SFGE into two engines: one for DFA rules and one for pathless rules.
-		return true;
-	}
-
-	private static processStderr(output: string): string {
```
Comment (jfeingold35) [inline src/lib/sfge/BaseSfgeEngine.ts:202]: No change. Just moved.

File: src/lib/sfge/SfgeWrapper.ts [hunk 8]
```diff
@@ -164,27 +150,89 @@ export class SfgeWrapper extends CommandLineSupport {
...
+	protected async execute(): Promise<string> {
 		return super.runCommand();
 	}
+}
 
-	public static async getCatalog() {
-		const wrapper = await SfgeWrapper.create({
-			targets: [],
-			projectDirs: [],
+export class SfgeCatalogWrapper extends SfgeWrapper {
```
Comment (jfeingold35) [inline src/lib/sfge/SfgeWrapper.ts:166]: Separated the wrapper into catalog wrapper and execution wrapper, analogous to PMD. Most of the behavior is the same, though.

File: test/commands/scanner/run.severity.test.ts [hunk 9]
```diff
@@ -5,7 +5,7 @@ import path = require('path');
 
 
 Messages.importMessagesDirectory(__dirname);
-const runMessages = Messages.loadMessages('@salesforce/sfdx-scanner', 'run');
+const runOutputProcessorMessages = Messages.loadMessages('@salesforce/sfdx-scanner', 'RunOutputProcessor');
```
Comment (jfeingold35) [inline test/commands/scanner/run.severity.test.ts:8]: Variable renaming.

File: test/commands/scanner/run.test.ts [hunk 10]
```diff
@@ -715,7 +716,14 @@ describe('scanner:run', function () {
 				expect(ctx.stdout).to.contain(runMessages.getMessage('output.filtersIgnoredCustom'));
 			});
 
-
+		setupCommandTest
```
Comment (jfeingold35) [inline test/commands/scanner/run.test.ts:719]: Aside from the variable renaming, I had to add a bit more test coverage since we dropped below the acceptable minimum.

File: test/lib/Controller.test.ts [hunk 11]
```diff
@@ -17,7 +17,7 @@ describe('Controller.ts tests', () => {
 		const engines: RuleEngine[] = await Controller.getAllEngines();
 		const names: string[] = engines.map(e => e.getName());
 
-		expect(engines.length, names + '').to.equal(9);
+		expect(engines.length, names + '').to.equal(10);
```
Comment (jfeingold35) [inline test/lib/Controller.test.ts:20]: New engine means a bunch of incrementing here.

File: test/lib/RuleManager.test.ts [hunk 12]
```diff
@@ -257,7 +257,7 @@ describe('RuleManager', () => {
 						parsedRes = JSON.parse(results);
 					}
 					expect(parsedRes).to.be.an("array").that.has.length(1);
-					Sinon.assert.callCount(uxSpy, 0);
+					Sinon.assert.callCount(uxSpy, 1);
```
Comment (jfeingold35) [inline test/lib/RuleManager.test.ts:260]: None of these calls have the `--projectdir`, so they're all trying to log the warning about skipping SFGE. Hence, the incrementation by one.

File: test/lib/sfge/SfgePathlessEngine.test.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+			(testEngine as any).catalog = fakeCatalog;
+			runSfgeStub = Sinon.stub(SfgeExecutionWrapper, 'runSfge').resolves("VIOLATIONS_START[]VIOLATIONS_END");
+		});
+
+		afterEach(() => {
+			Sinon.restore();
+		});
+
+
+		it('Only attempts to run rules in catalog', async () => {
```
Comment (jfeingold35) [inline test/lib/sfge/SfgePathlessEngine.test.ts:135]: Testing that the rule filtering works properly.

File: test/lib/sfge/SfgeDfaEngine.test.ts [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+			(testEngine as any).catalog = fakeCatalog;
+			runSfgeStub = Sinon.stub(SfgeExecutionWrapper, 'runSfge').resolves("VIOLATIONS_START[]VIOLATIONS_END");
+		});
+
+		afterEach(() => {
+			Sinon.restore();
+		});
+
+
+		it('Only attempts to run rules in catalog', async () => {
```
Comment (jfeingold35) [inline test/lib/sfge/SfgeDfaEngine.test.ts:135]: Testing that the rule filtering works properly.

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 15]
```diff
@@ -88,7 +94,7 @@ export class SfgeEngine extends AbstractRuleEngine {
 				sourcepackage: "sfge",
 				name,
 				description,
-				// SFGE rules each belong to exactly one category, so the string must be converted to a singleton array.
+				// SFGE rules each belogn to exactly one category, so the string must be converted to a singleton array.
```
Comment (rmohan20) [inline src/lib/sfge/BaseSfgeEngine.ts:97]: looks like an unintentional change

File: src/lib/sfge/BaseSfgeEngine.ts [hunk 16]
```diff
@@ -32,54 +35,57 @@ type SfgeViolation = {
...
+	 * @override
 	 */
 	public async getCatalog(): Promise<Catalog> {
-		const catalogOutput: string = await SfgeWrapper.getCatalog();
+		// If we've already generated the catalog, don't generate another.
+		if (this.catalog) {
+			return this.catalog;
+		}
+		// DFA engine should catalog DFA rules, and non-DFA engine should catalog non-DFA rules.
+		const ruleType = this.isDfaEngine() ? "dfa" : "pathless";
```
[the remark this answers is not in the document]
Comment (rmohan20) [reply src/lib/sfge/BaseSfgeEngine.ts:74]: Can we make the strings constants in the module?

File: src/lib/sfge/BaseSfgeEngine.ts (hunk 1 above)
Comment (rmohan20) [reply src/lib/sfge/BaseSfgeEngine.ts:133]: I like the intention for the distinction, but wondering if the inconsistency could be more confusing. 
If we were to compare this with eslint-typescript which needs ts-config parameter in most cases (unless there's a tsconfig in the current directory), wouldn't we throw the error of missing tsconfig even if eslint-typescript is enabled by default?
Since sfge would be disabled by default, and the user needs to intentionally set it to enabled, I feel it's fair to be consistent and throw the error if --projectdir is missing no matter how it was invoked.

Comment (rmohan20) [reply src/lib/sfge/BaseSfgeEngine.ts:133]: We could postpone the check as far as to see if we have files matching the pattern sfge looks for. I believe that's how the check works for eslint-typescript as well. If no files match, SFCA already knows that this engine will not be run, and no point in checking for missing parameters.

**Comment by jfeingold35** [comment]
This PR has been superseded by #786 .
